### PR TITLE
[core] add Data convertible on ios

### PR DESCRIPTION
**Pull request by Kudo**
# Why

to support Blob data type for expo-sqlite/next. This pr adds `Data <-> Uint8Array` convertible on ios

# How

- fix dict values does not convert correctly
- add Data <-> Uint8Array convertible
  - why not using TypedArray?
    since hermes requires accessing everything in the js thread, i think our current TypedArray implementation may violate the hermes restriction on AsyncFunction. the current `DynamicDictionaryType` will convert UintArray to Data in js thread because jumping to async function queue thread.
- it's best to align the conversion between function input and output. adding Data -> Uint8Array conversion in `convertObjCObjectToJSIValue`
- adding call invoker for async js unit test. the code is referenced from the expo-modules-core.

# Test Plan

- add `BlobConvertiblesSpec` unit test

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by amandeepmittal — APPROVED** [review]

File: packages/expo-modules-core/ios/Arguments/AnyArgument.swift [hunk 1]
```diff
@@ -33,14 +33,24 @@ extension Double: AnyArgument {}
...
+  }
+}
 
 extension Optional: AnyArgument where Wrapped: AnyArgument {
   public static func getDynamicType() -> AnyDynamicType {
     return DynamicOptionalType(wrappedType: ~Wrapped.self)
   }
 }
 
+extension Dictionary: AnyArgument {
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Arguments/AnyArgument.swift:48]: What do you think about adding a constraint here?

```suggestion
extension Dictionary: AnyArgument where Key: Hashable {
```

File: packages/expo-modules-core/ios/Arguments/AnyArgument.swift [hunk 2]
```diff
@@ -33,14 +33,24 @@ extension Double: AnyArgument {}
 extension CGFloat: AnyArgument {}
 
 extension String: AnyArgument {}
-extension Dictionary: AnyArgument {}
+extension Data: AnyArgument {
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/Arguments/AnyArgument.swift:36]: Nitpick: so far the extensions were ordered by "complexity", from primitive types to arrays, optionals, dictionaries – can you put it at the end of the file as it's a less common type?

File: packages/expo-modules-core/ios/DynamicTypes/DynamicDictionaryType.swift [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+  func cast<ValueType>(_ value: ValueType, appContext: AppContext) throws -> Any {
+    if let value = value as? [AnyHashable: Any] {
+      return try value.mapValues { try valueType.cast($0, appContext: appContext) }
+    }
+    throw Conversions.CastingException<[AnyHashable: Any]>(value)
+  }
+
+  var description: String {
+    "[AnyHashable: \(valueType.description)]"
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/DynamicTypes/DynamicDictionaryType.swift:43]: Probably we should also pass a key type to this dynamic type so the description could be more accurate. If not yet, at least let's use Hashable instead (`AnyHashable` is a struct, `Hashable` is a protocol which seems better in my opinion)
```suggestion
    "[Hashable: \(valueType.description)]"
```

File: packages/expo-modules-core/ios/DynamicTypes/DynamicFoundationDataType.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
+// Copyright 2015-present 650 Industries. All rights reserved.
+
+/**
+ A dynamic type representing Swift `Data` or Objective-C `NSData` type and backing by JavaScript `Uint8Array`.
+ */
+internal struct DynamicFoundationDataType: AnyDynamicType {
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/DynamicTypes/DynamicFoundationDataType.swift:6]: `DynamicDataType` would be fine as well 😉

File: packages/expo-modules-core/ios/JSI/EXJSIConversions.mm [hunk 5]
```diff
@@ -55,6 +55,17 @@
   return result;
 }
 
+jsi::Value createUint8Array(jsi::Runtime &runtime, NSData *data) {
+  auto arrayBufferCtor = runtime.global().getPropertyAsFunction(runtime, "ArrayBuffer");
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/JSI/EXJSIConversions.mm:59]: You're using `runtime.global()` twice, let's use a separate variable for it 😉

File: packages/expo-modules-core/ios/JSI/EXJSIConversions.mm [hunk 6]
```diff
@@ -73,6 +84,8 @@
       return convertNSNumberToJSIBoolean(runtime, (NSNumber *)value);
     }
     return convertNSNumberToJSINumber(runtime, (NSNumber *)value);
+  } else if ([value isKindOfClass:[NSData class]]) {
```
Comment (tsapeta) [inline packages/expo-modules-core/ios/JSI/EXJSIConversions.mm:87]: `NSData` is less common than `NSDictionary` or `NSArray`, so I would place it after them (small optimizations 🤷‍♂️)

File: packages/expo-modules-core/ios/DynamicTypes/DynamicDictionaryType.swift (hunk 3 above)
Comment (Kudo) [reply packages/expo-modules-core/ios/DynamicTypes/DynamicDictionaryType.swift]: thanks! good to learn more about swift hashable.
i would still not include key type if the key type is only for description.

**Review by amandeepmittal — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]

**Pull request merged**